src/shims/windows/handle.rs

```diff
@@ -0,0 +1,162 @@
+            Self::THREAD_DISCRIMINANT => Some(Self::Thread(data.into())),
+            _ => None,
+        }
+    }
+
+    /// see docs for `to_packed`
+    fn from_packed(handle: i32) -> Option<Self> {
+        let handle_bits = handle as u32;
+
+        let disc_size = Self::packed_disc_size();
+        let data_size = Self::USABLE_BITS - disc_size;
+
+        // the lower `data_size` bits of this mask are 1
+        let data_mask = 2u32.pow(data_size) - 1;
+
+        // the discriminant is stored right above the lower `data_size` bits
+        let discriminant = handle_bits >> data_size;
+
+        // the data is stored in the lower `data_size` bits
+        let data = handle_bits & data_mask;
+
+        Self::new(discriminant, data)
+    }
+}
+
+/// Miri representation of a Windows `HANDLE`
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Handle {
+    Null, // = 0
+
+    // pseudo-handles
+    // The lowest pseudo-handle is -6, so miri pseduo-handles start at -7 to break code hardcoding these values
+    CurrentThread, // = -7
+
+    // real handles
+    Thread(ThreadId),
+}
+
+impl Handle {
+    const CURRENT_THREAD_VALUE: i32 = -7;
+
+    fn to_packed(self) -> i32 {
+        match self {
+            Self::Null => 0,
+            Self::CurrentThread => Self::CURRENT_THREAD_VALUE,
+            Self::Thread(thread) => RealHandle::Thread(thread).to_packed(),
+        }
+    }
+
+    pub fn to_scalar(self, cx: &impl HasDataLayout) -> Scalar<Tag> {
+        // 64-bit handles are sign extended 32-bit handles
+        // see https://docs.microsoft.com/en-us/windows/win32/winprog64/interprocess-communication
+        let handle = self.to_packed().into();
+
+        Scalar::from_machine_isize(handle, cx)
+    }
+
+    fn from_packed(handle: i64) -> Option<Self> {
+        let current_thread_val = Self::CURRENT_THREAD_VALUE as i64;
+
+        if handle == 0 {
+            Some(Self::Null)
+        } else if handle == current_thread_val {
+            Some(Self::CurrentThread)
+        } else if let Ok(handle) = handle.try_into() {
+            match RealHandle::from_packed(handle)? {
+                RealHandle::Thread(id) => Some(Self::Thread(id)),
+            }
+        } else {
+            // if a handle doesn't fit in an i32, it isn't valid.
+            None
+        }
+    }
+
+    pub fn from_scalar<'tcx>(
+        handle: Scalar<Tag>,
+        cx: &impl HasDataLayout,
+    ) -> InterpResult<'tcx, Option<Self>> {
+        let handle = handle.to_machine_isize(cx)?;
+
+        Ok(Self::from_packed(handle))
+    }
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+
+#[allow(non_snake_case)]
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn CloseHandle(&mut self, handle_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        match Handle::from_scalar(this.read_scalar(handle_op)?.check_init()?, this)? {
```

Comment:
Please let-bind `this.read_scalar(handle_op)?.check_init()?` for more readability (also following the usual style: shims start by reading all their arguments)